File: library/src/main/java/com/benny/library/neobinding/bind/ViewCreator.java
package com.benny.library.neobinding.bind;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

/**
 * Created by benny on 11/5/15.
 */

public class ViewCreator<T> {
    private ViewModelBinder<T> viewModelBinder;
    private int layoutId;

    public static <T> ViewCreator<T> create(ViewModelBinder<T> viewModelBinder, int layoutId) {
        return new ViewCreator<>(viewModelBinder, layoutId);
    }

    public ViewCreator() {
    }

    private ViewCreator(ViewModelBinder<T> viewModelBinder, int layoutId) {
        this.viewModelBinder = viewModelBinder;
        this.layoutId = layoutId;
    }

    public View view(ViewGroup container) {
        View view = LayoutInflater.from(container.getContext()).inflate(layoutId, container, false);
        BindableModel<T> bindableModel = viewModelBinder.bind(view);
        view.setTag(bindableModel);
        return view;
    }
}

File: library/src/main/kotlin/com/benny/library/neobinding/kotlin/bind/ViewCreator.kt
package com.benny.library.neobinding.kotlin.bind

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.benny.library.neobinding.kotlin.view.ViewComponent

/**
 * Created by benny on 11/18/15.
 */

public open class ViewCreator<T> : IViewCreator<T> {
    val viewComponent: ViewComponent
    val viewBinder: ViewBinder<BindableModel<in T> >

    constructor(viewBinder: ViewBinder<BindableModel<in T> >, viewComponent: ViewComponent) {
        this.viewBinder = viewBinder
        this.viewComponent = viewComponent
    }

    override fun view(context: Context, container: ViewGroup?): View {
        val view = viewComponent.createView(context)
        view.tag = viewBinder.bind(view)
        return view
    }

    override fun viewTypeFor(data: T?, position: Int): Int {
        return  0
    }

    override fun viewTypeCount(): Int {
        return 1
    }
}